consensus/istanbul/backend/internal/enodes/val_enode_db.go

```diff
@@ -212,78 +213,122 @@ func (vet *ValidatorEnodeDB) String() string {
 	vet.lock.RLock()
 	defer vet.lock.RUnlock()
 	entry, err := vet.getAddressEntry(address)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
-	return entry.enodeURL, nil
+	return entry.Node, nil
 }
 
-// GetAddressFromEnodeURL will return the address for an enodeURL if it's known
-func (vet *ValidatorEnodeDB) GetAddressFromEnodeURL(enodeURL string) (common.Address, error) {
+// GetViewFromAddress will return the view for an address if it's known
+func (vet *ValidatorEnodeDB) GetViewFromAddress(address common.Address) (*istanbul.View, error) {
 	vet.lock.RLock()
 	defer vet.lock.RUnlock()
-	return vet.getAddressFromEnodeURL(enodeURL)
+	entry, err := vet.getAddressEntry(address)
+	if err != nil {
+		return nil, err
+	}
+	return entry.View, nil
 }
 
-func (vet *ValidatorEnodeDB) getAddressFromEnodeURL(enodeURL string) (common.Address, error) {
-	rawEntry, err := vet.db.Get(enodeURLKey(enodeURL), nil)
+// GetAddressFromNodeID will return the address for an nodeID if it's known
+func (vet *ValidatorEnodeDB) GetAddressFromNodeID(nodeID enode.ID) (common.Address, error) {
+	vet.lock.RLock()
+	defer vet.lock.RUnlock()
+	return vet.getAddressFromNodeID(nodeID)
+}
+
+func (vet *ValidatorEnodeDB) getAddressFromNodeID(nodeID enode.ID) (common.Address, error) {
+	rawEntry, err := vet.db.Get(nodeIDKey(nodeID), nil)
 	if err != nil {
 		return common.ZeroAddress, err
 	}
 	return common.BytesToAddress(rawEntry), nil
 }
 
+func (vet *ValidatorEnodeDB) GetAllValEnodes() (map[common.Address]*AddressEntry, error) {
+	vet.lock.RLock()
+	defer vet.lock.RUnlock()
+	var entries = make(map[common.Address]*AddressEntry)
+
+	err := vet.iterateOverAddressEntries(func(address common.Address, entry *AddressEntry) error {
+		entries[address] = entry
+		return nil
+	})
+
+	if err != nil {
+		vet.logger.Error("ValidatorEnodeDB.GetAllAddressEntries error", "err", err)
+		return nil, err
+	}
+
+	return entries, nil
+}
+
 // Upsert will update or insert a validator enode entry; given that the existing entry
 // is older (determined by view parameter) that the new one
-func (vet *ValidatorEnodeDB) Upsert(remoteAddress common.Address, enodeURL string, view *istanbul.View) error {
+func (vet *ValidatorEnodeDB) Upsert(valEnodeEntries map[common.Address]*AddressEntry) error {
 	vet.lock.Lock()
 	defer vet.lock.Unlock()
 
-	currentEntry, err := vet.getAddressEntry(remoteAddress)
-	isNew := err == leveldb.ErrNotFound
+	batch := new(leveldb.Batch)
+	peersToRemove := make([]*enode.Node, 0, len(valEnodeEntries))
+	peersToAdd := make(map[common.Address]*enode.Node)
 
-	// Check errors
-	if !isNew && err != nil {
-		return err
-	}
+	for remoteAddress, addressEntry := range valEnodeEntries {
+		currentEntry, err := vet.getAddressEntry(remoteAddress)
+		isNew := err == leveldb.ErrNotFound
 
-	// If it is an old message, ignore it.
-	if err == nil && view.Cmp(currentEntry.view) <= 0 {
-		return errOldAnnounceMessage
-	}
+		// Check errors
+		if !isNew && err != nil {
+			return err
```

Comment:
Should we return an error? Or continue?

Reply:
The thought is that `Upsert` will behave like a transaction.  All entries must be successful, or none will be.